pkg/nimble/autoconn/nimble_autoconn.c

```diff
@@ -0,0 +1,301 @@
+{
+    bluetil_ad_data_t incomp;
+    if (bluetil_ad_find(ad, BLE_GAP_AD_UUID16_INCOMP, &incomp) == BLUETIL_AD_OK) {
+        uint16_t filter_uuid = SVC_FILTER;
+        for (unsigned i = 0; i < incomp.len; i += 2) {
+            if (memcmp(&filter_uuid, &incomp.data[i], 2) == 0) {
+                return 1;
+            }
+        }
+    }
+
+    return 0;
+}
+
+static void _on_scan_evt(uint8_t type, const ble_addr_t *addr, int8_t rssi,
+                         const uint8_t *ad_buf, size_t ad_len)
+{
+    (void)rssi;
+
+    /* we are only interested in ADV_IND packets, the rest can be dropped right
+     * away */
+    if (type != BLE_HCI_ADV_TYPE_ADV_IND) {
+        return;
+    }
+
+    bluetil_ad_t ad = {
+        .buf  = (uint8_t *)ad_buf,
+        .pos  = ad_len,
+        .size = ad_len
+    };
+
+    /* for connection checking we need the address in network byte order */
+    uint8_t addrn[BLE_ADDR_LEN];
+    bluetil_addr_swapped_cp(addr->val, addrn);
+
+    if (_filter_uuid(&ad) && !nimble_netif_conn_connected(addrn)) {
+        nimble_autoconn_disable();
+        _state = STATE_CONN;
+        int res = nimble_netif_connect(addr, &_conn_params, _conn_timeout);
+        assert(res >= 0);
+        (void)res;
+        DEBUG("[autoconn] SCAN success, initiating connection\n");
+    }
+}
+
+static void _on_netif_evt(int handle, nimble_netif_event_t event)
+{
+    switch (event) {
+        case NIMBLE_NETIF_CONNECTED_MASTER:
+            DEBUG("[autoconn] CONNECTED as master %i\n", handle);
+            assert(_state == STATE_CONN);
+            _state = STATE_IDLE;
+            nimble_autoconn_enable();
+            break;
+        case NIMBLE_NETIF_CONNECTED_SLAVE:
+            DEBUG("[autoconn] CONNECTED as slave %i\n", handle);
+            nimble_autoconn_enable();
+            break;
+        case NIMBLE_NETIF_CLOSED_MASTER:
+            DEBUG("[autoconn] CLOSED master connection\n");
+            nimble_autoconn_enable();
+            break;
+        case NIMBLE_NETIF_CLOSED_SLAVE:
+            DEBUG("[autoconn] CLOSED slave connection\n");
+            nimble_autoconn_enable();
+            break;
+        case NIMBLE_NETIF_CONNECT_ABORT:
+            DEBUG("[autoconn] CONNECT ABORT\n");
+            assert(_state == STATE_CONN);
+            _state = STATE_IDLE;
+            nimble_autoconn_enable();
+            break;
+        case NIMBLE_NETIF_CONN_UPDATED:
+            DEBUG("[autoconn] CONNECTION UPDATED %i\n", handle);
+            /* nothing to do here */
+            break;
+        default:
+            /* this should never happen */
+            assert(0);
+    }
+}
+
+static int _conn_update(nimble_netif_conn_t *conn, int handle, void *arg)
+{
+    (void)conn;
+    nimble_netif_update(handle, (const struct ble_gap_upd_params *)arg);
+    return 0;
+}
+
+int nimble_autoconn_init(const nimble_autoconn_params_t *params,
+                         const uint8_t *ad, size_t adlen)
+{
+    /* register our event callback */
```

Comment:
Is it possible to provide an assert if netif is not initialized?

Reply:
No so easily without. But I don't think its really needed, as autoconn has nimble_netif as its dependency, and so nimble_netif is always initialized if autoconn is included in the build.

That is, as long auto_init is used. And not using auto_init means that people have to carefully initialized everything themselves, and so they have to worry about these kind of dependencies themselves :-)